### PR TITLE
Proposed changes to action_loop

### PR DESCRIPTION
**Pull request by bruceadams**
The key change here is using more of the power of Serde for parsing the input.

There are several other, bigger picture things that should get discussed elsewhere.

### CODE REVIEW

**Review by dubee — COMMENTED** [review]
@sciabarracom, can you review this one?

**Comment by sciabarracom** [comment]
I am asking Roberto Diaz opinion as he is the original author of the Go ActionLoop

File: rust1.32/src/action_loop/src/main.rs [hunk 1]
```diff
@@ -1,49 +1,46 @@
...
+    env,
+    fs::File,
+    io::{stderr, stdin, stdout, BufRead, Write},
+    os::unix::io::FromRawFd,
+};
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+struct Input {
+    value: HashMap<String, Value>,
+    #[serde(flatten)]
```
Comment (rdiaz82) [inline rust1.32/src/action_loop/src/main.rs:16]: Awesome contribution, I didn't know about this!!

File: rust1.32/src/action_loop/src/main.rs [hunk 2]
```diff
@@ -1,49 +1,46 @@
...
-                            Ok(value) => payload = value,
-                            Err(err) => {
-                                eprintln!("Error parsing value json: {}", err);
-                                continue
-                            }
-                        }
-                    } else {
-                        env::set_var(format!("__OW_{}", key.to_uppercase()), val.to_string());
+            Ok(input) => {
+                for (key, val) in input.environment {
```
Comment (rdiaz82) [inline rust1.32/src/action_loop/src/main.rs:28]: What about this?
```
                input.environment.iter().for_each(|(key, val)|  env::set_var(format!("__OW_{}", key.to_uppercase()), val.to_string()));
```

**Review by rdiaz82 — APPROVED** [review]
LGTM! just small comment
On the other hand, I think that we should add Cargo.lock to .gitignore

**Comment by bruceadams** [comment]
My understanding it that Rust projects that produce a binary _should_ check `Cargo.lock` into source control and Rust projects that provide a library for others to use should **not** check in `Cargo.lock`. I wonder if I can find a write up for that?

File: rust1.32/src/action_loop/src/main.rs (hunk 2 above)
Comment (bruceadams) [reply rust1.32/src/action_loop/src/main.rs:28]: The `iter()` approach is fine. I suspect the generated code is essentially the same for either approach.

**Comment by bruceadams** [comment]
Ah, yup. Right there in the documentation https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html

> If you’re building a library that other packages will depend on, put `Cargo.lock` in your `.gitignore`. If you’re building an executable like a command-line tool or an application, check `Cargo.lock` into `git`. If you're curious about why that is, see ["Why do binaries have `Cargo.lock` in version control, but not libraries?" in the FAQ.](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries)

File: rust1.32/src/action_loop/src/main.rs (hunk 2 above)
Comment (bruceadams) [reply rust1.32/src/action_loop/src/main.rs:28]: Running the `iter()` approach through `cargo fmt` gives this:

```rust
               input.environment.iter().for_each(|(key, val)| {
                   env::set_var(format!("__OW_{}", key.to_uppercase()), val.to_string())
               });
```

I find the `for` slightly easier to read 🤷.

**Comment by sciabarracom** [comment]
Is it possible to get a fix also for the example?

File: rust1.32/src/action_loop/src/main.rs (hunk 2 above)
Comment (rdiaz82) [reply rust1.32/src/action_loop/src/main.rs:28]: yeah! this is just a personal style suggestion, feel free to ignore it 😄

**Comment by rdiaz82** [comment]
I would like to see this method signature `fn main(input_data: Value) -> Result<Value, Error>` I have tested locally and it works nice

**Comment by sciabarracom** [comment]
Which JSON the mainloop will emit if the result is an error?

**Comment by rdiaz82** [comment]
we can create one specific for that, what about `{"result": "error generating output response"}` or something like that :construction_worker:

**Comment by rdiaz82** [comment]
I think that we should move this conversation to other place, to not spam the PR. I think that we can merge it and continue from that.

**Comment by sciabarracom** [comment]
Ok give me the time to check if it builds  and I will merge

**Pull request merged**